File: 12.clj

(defn simple-cost [items]
    (reduce +
        (map #(* (get % :price) (get % :count)) items)))

(defn create-price-between-cost [next min max percent]
    (fn [items]
        (-
            (next items)
            (reduce +
                (map #(* (* (/ percent 100) (:price %)) (:count %))
                    (filter #(and (<= min (:price %)) (<= (:price %) max)) items))))))

(defn create-month-cost [next day needle percent]
    (fn [items]
        (if (= day needle)
            (* (- 1 (/ percent 100)) (next items))
            (next items))))

(defn create-big-cost [next limit percent]
    (fn [items]
        (let [cost (next items)]
            (if (>= cost limit)
                (* (- 1 (/ percent 100)) cost)
                cost))))

(def day (. (java.time.LocalDate/now) getDayOfMonth))

(def price-between-cost (create-price-between-cost simple-cost 100 150 10))
(def month-cost (create-month-cost price-between-cost day 15 5))
(def cost (create-big-cost month-cost 1000 7))

(def items [
    {:count 2 :price 75}
    {:count 5 :price 150}
])

(println (str (int (cost items))))

;; java -cp clojure.jar clojure.main 12.clj

File: 12.php
<?php

function createSimpleCost()
{
    return function (array $items) {
        $itemCost = function ($item) {
            return $item['price'] * $item['count'];
        };

        return array_sum(
            array_map($itemCost, $items));
    };
}

function createPriceBetweenCost(callable $next, $min, $max, $percent)
{
    return function (array $items) use ($next, $min, $max, $percent) {
        $discount = function ($item) use ($percent) { return ($percent / 100) * $item['price'] * $item['count']; };
        $filter = function ($item) use ($min, $max) { return $min <= $item['price'] && $item['price'] <= $max; };

        return $next($items) - array_sum(
            array_map($discount,
                array_filter($items, $filter)));
    };
}

function createMonthCost(callable $next, $day, $needle, $percent)
{
    return function (array $items) use ($next, $day, $needle, $percent) {
        return ($day == $needle) ? (1 - $percent / 100) * $next($items) : $next($items);
    };
}

function createBigCost(callable $next, $limit, $percent)
{
    return function (array $items) use ($next, $limit, $percent) {
        $cost = $next($items);
        return ($cost >= $limit) ? (1 - $percent / 100) * $cost : $cost;
    };
}

##################################

$simpleCost = createSimpleCost();
$priceBetweenCost = createPriceBetweenCost($simpleCost, 100, 150, 9);
$monthCost = createMonthCost($priceBetweenCost, date('d'), 15, 5);
$cost = createBigCost($monthCost, 1000, 7);

##################################

$items = [
    ['count' => 2, 'price' => 75],
    ['count' => 5, 'price' => 150],
];

echo $cost($items) . PHP_EOL;